src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java

```diff
@@ -111,6 +111,28 @@ public static void check(final HaxeVarDeclaration var, final AnnotationHolder ho
     if (field.hasInitializer() && field.hasTypeTag()) {
       TypeTagChecker.check(field.getPsi(), field.getTypeTagPsi(), field.getInitializerPsi(), false, holder);
     }
+
+    // Checking for variable redefinition.
+    HashSet<HaxeClassModel> classSet = new HashSet<HaxeClassModel>();
+    HaxeClassModel fieldDeclaringClass = field.getDeclaringClass();
+    classSet.add(fieldDeclaringClass);
+    while (fieldDeclaringClass != null) {
+      fieldDeclaringClass = fieldDeclaringClass.getParentClass();
+      if (classSet.contains(fieldDeclaringClass)) {
+        break;
+      } else {
+        classSet.add(fieldDeclaringClass);
+      }
+      if (fieldDeclaringClass != null) {
+        for (HaxeFieldModel parentField : fieldDeclaringClass.getFields()) {
+          if (parentField.getName().equals(field.getName())) {
+            holder.createErrorAnnotation(field.getDeclarationPsi(), "Redefinition of variable '" + field.getName()
+              + "' in subclass is not allowed. Previously declared at '" + fieldDeclaringClass.getName() + "'.");
+            break;
```

Comment:
I see what you're saying, but I think that it's OK to use the models to traverse, since the whole Semantic Annotation system is built on the models.  It's better to not have a mixed metaphor.